internal/telemetry/collector/segmentio.go

```diff
@@ -0,0 +1,469 @@
+		crCl:            crClient,
+		systemNamespace: systemNamespace,
+		childScheme:     s,
+		concurrency:     concurrency,
+		childFactory:    defaultClientFactory,
+	}, nil
+}
+
+func (t *SegmentIO) Collect(ctx context.Context) error {
+	l := ctrl.LoggerFrom(ctx)
+
+	mgmt := new(kcmv1.Management)
+	if err := t.retry(ctx, func() error {
+		return t.crCl.Get(ctx, client.ObjectKey{Name: kcmv1.ManagementName}, mgmt)
+	}); err != nil {
+		return fmt.Errorf("failed to get Management: %w", err)
+	}
+
+	mgmtID := string(mgmt.UID)
+
+	templatesList := new(kcmv1.ClusterTemplateList)
+	if err := t.retry(ctx, func() error {
+		return t.crCl.List(ctx, templatesList, client.InNamespace(t.systemNamespace))
+	}); err != nil {
+		return fmt.Errorf("failed to list ClusterTemplates: %w", err)
+	}
+
+	tplName2Template := make(map[string]*kcmv1.ClusterTemplate, len(templatesList.Items))
+	for _, template := range templatesList.Items {
+		tplName2Template[template.Name] = &template
+	}
+
+	clusterDeployments := new(kcmv1.ClusterDeploymentList)
+	if err := t.retry(ctx, func() error {
+		return t.crCl.List(ctx, clusterDeployments)
+	}); err != nil {
+		return fmt.Errorf("failed to list ClusterDeployments: %w", err)
+	}
+
+	mcsList := new(kcmv1.MultiClusterServiceList)
+	if err := t.retry(ctx, func() error {
+		return t.crCl.List(ctx, mcsList)
+	}); err != nil {
+		return fmt.Errorf("failed to list MultiClusterServices: %w", err)
+	}
+
+	var partialClustersToMatch []metav1.PartialObjectMetadata
+	if err := t.retry(ctx, func() error {
+		var lerr error
+		partialClustersToMatch, lerr = t.getPartialClustersToCountServices(ctx)
+		return lerr
+	}); err != nil {
+		return fmt.Errorf("failed to list clusters to count services from: %w", err)
+	}
+
+	var (
+		wg  sync.WaitGroup
+		sem = make(chan struct{}, t.concurrency)
+	)
+
+	// TODO: NOTE: should we also retry each of the clusters or is it okay to skip on failure?
+	for _, cld := range clusterDeployments.Items {
+		wg.Add(1)
+		sem <- struct{}{}
+
+		go func() {
+			defer func() {
+				wg.Done()
+				<-sem
+			}()
+			ll := l.WithValues("cld", cld.Namespace+"/"+cld.Name)
+
+			template, ok := tplName2Template[cld.Spec.Template] // NOTE: it's okay to read concurrently here since NO writes occure
+			if !ok {
+				template = &kcmv1.ClusterTemplate{}
+			}
+
+			childCl, err := getChildClient(ctx, t.crCl, cld.Name, t.systemNamespace, t.childScheme, t.childFactory)
+			if err != nil {
+				ll.Error(err, "failed to get child kubeconfig")
+				return
+			}
+
+			props, err := t.collectChildProperties(ctx, childCl)
+			if err != nil {
+				ll.Error(err, "failed to collect properties")
+				props = make(map[string]any) // avoid runtime error
+			}
+
+			// this map is shared within a goroutine, so no sync is required
+			props["cluster"] = cld.Namespace + "/" + cld.Name
+			props["clusterDeploymentID"] = string(cld.UID)
+			props["clusterID"] = "" // TODO: get k0s cluster ID once it's exposed in k0smotron API
```

Comment:
nit: k0smotron already exposes the k0s cluster ID as an annotation of the cluster object: https://github.com/k0sproject/k0smotron/issues/683

but we can implement it in a separate PR

Reply:
Thanks for pointing that, done